### PR TITLE
Contour domain

### LINKED ISSUE #571: Optimize particles on n-D contours
This entails implementing a ContourDomain with necessary APIs to interact with the optimizer.

### LINKED ISSUE #572: n-D contours I/O
This entails reading contours (polylines) as input for shapeworks optimize and saving particle systems (optimized and checkpoints) for contour domains (this existing write functions might be reused for this purpose);

### PR DESCRIPTION
**Pull request by medakk**
* Allows us to run shapeworks on contours. These are VTK .vtp files, and utility functions for writing to this format are in ~~GroomUtils.py for now~~ ShapeCohortGen.CohortGenUtils.py for now. 
* Most of the changes(geodesics, geodesic walk, clockwise/anticlockwise splitting, etc) are in ContourDomain.h/cpp with some API changes in the optimizer:
  * ParticleDomain::Distance now takes particle indices, and optionally a vector to return gradient of distance
  * ParticleSurfaceNeighborhood now exposes an option to disable cosine-falloff weighting
  * ParticleDomain::GetPositionAfterSplit allows domains to return a consistent splitting direction.
These API changes should have no effect on image and mesh domain
* Adds a unit test for contour domain in OptimizeTests
* Adds a new use case supershapes_1mode_contour (TODO for @medakk upload this to data portal). Results on this:
20210403_supershapesworking

Use `python RunUseCase.py --use_case supershapes_1mode_contour` to test this use case
* Small change to ShapeCohortGen.Supershapes(@iyerkrithika21) to allow for generating 2D supershapes with parameters over chi square distribution

Closes #571
Closes #572

### CODE REVIEW

**Comment by sheryjoe** [comment]
@medakk can we reconstruct the contour? e.g. draw polylines between particles?

**Comment by medakk** [comment]
Not right now, no. There is no ordering enforced among the particle positions.

**Comment by cchriste** [comment]
Same story with the #1148, so I marked it as a Draft just to blur that merge button. We'll merge the release independently, then new pull requests after that.

**Comment by cchriste** [comment]
This PR has been updated with latest master-with-release-6 merged.
It should be good to go once the Windows build issue is fixed. Any ideas?

**Comment by medakk** [comment]
Supershapes are now generated with chi square distribution as discussed in #1177. The use case has been updated:

20210403_supershapesworking

**Comment by medakk** [comment]
Builds on all platforms now!

**Comment by sheryjoe** [comment]
> Allows us to run shapeworks on contours. These are VTK .vtp files, and utility functions for writing to this format are in GroomUtils.py for now ShapeCohortGen.CohortGenUtils.py for now.

These utils functions should be part of the shapeworks pymodule (e.g., shapeworks.io or shapeworks.utils)

> Small change to ShapeCohortGen.Supershapes(@iyerkrithika21) to allow for generating 2D supershapes with parameters over chi square distribution

Are 3D supershapes generated the same way? (i.e., parameters drawn from chi-square distribution). If not, please update the 3D code to be consistent with the 2d one.

**Comment by sheryjoe** [comment]
BTW, very nice correspondence model! 👍  @medakk

**Comment by medakk** [comment]
> These utils functions should be part of the shapeworks pymodule (e.g., shapeworks.io or shapeworks.utils)

I spoke to @iyerkrithika21 about this and that module is a work in progress right now. So it'll be in ShapeCohortGen and moved to the pymodule once that is ready

> Are 3D supershapes generated the same way? (i.e., parameters drawn from chi-square distribution). If not, please update the 3D code to be consistent with the 2d one.

No, they use uniform distribution. Its a small change to make, but I'm not sure where supershapes-3d are being used across our codebase right now in ordered to make sure all of that continues to function. @jadie1 ?

**Comment by sheryjoe** [comment]
> > These utils functions should be part of the shapeworks pymodule (e.g., shapeworks.io or shapeworks.utils)
> 
> I spoke to @iyerkrithika21 about this and that module is a work in progress right now. So it'll be in ShapeCohortGen and moved to the pymodule once that is ready
> 

Ok. Is there an issue for this?

**Comment by medakk** [comment]
> Ok. Is there an issue for this?

I created #1195

**Comment by jadie1** [comment]
> No, they use uniform distribution. Its a small change to make, but I'm not sure where supershapes-3d are being used across our codebase right now in order to make sure all of that continues to function. @jadie1 ?

@medakk The only place the 3d supershapes code would need to be changed is here: [Supershapes.py Lines 19 and 20](https://github.com/SCIInstitute/ShapeWorks/blob/master/Python/ShapeCohortGenPackage/ShapeCohortGen/Supershapes.py#L19,L20)
And the only place the 3D supershapes are currently used is in the shape generation notebook, so we would just want to run that to verify that changing the sampling distribution doesn't break anything. Let me know if you'd like me to make the change.

**Comment by medakk** [comment]
@jadie1 It'd be great if you could make that change. thanks!

**Comment by jadie1** [comment]
I updated the 3D supershapes to use Chi squared with 4 degrees of freedom and made an issue to add segs and image generation for 2D supershapes #1197

**Comment by medakk** [comment]
Thanks @jadie1 ! I've update the supershapes 2D generation according to use a class extended from the CohortGenerator. Let me know if I need to make any more changes here

File: Libs/Optimize/ParticleSystem/ContourDomain.cpp [hunk 1]
```diff
@@ -0,0 +1,348 @@
+#include "ContourDomain.h"
+#include <iostream>
+#include <fstream>
+#include <vtkDijkstraGraphGeodesicPath.h>
+#include <vtkDoubleArray.h>
+
+
+namespace itk{
```
Comment (akenmorris) [inline Libs/Optimize/ParticleSystem/ContourDomain.cpp:8]: 😭

**Comment by akenmorris** [comment]
Code looks good.  I tried it out and it works great.

Anything else need to be done before merging?

File: Libs/Optimize/ParticleSystem/ContourDomain.cpp (hunk 1 above)
Comment (medakk) [reply Libs/Optimize/ParticleSystem/ContourDomain.cpp:8]: :( needed for the itkSmartPointer stuff

**Comment by medakk** [comment]
Not that I know of, I think its good to merge.

File: Libs/Optimize/ParticleSystem/ContourDomain.cpp (hunk 1 above)
Comment (akenmorris) [reply Libs/Optimize/ParticleSystem/ContourDomain.cpp:8]: Yeah, I know it's a subclass of ParticleDomain

**Pull request merged**